posts/2022-10-04-microprofile-serverless-ibm-code-engine.adoc

```diff
@@ -0,0 +1,412 @@
+
+When run within the context of Open Liberty, the `CloudEventsProvider` class is automatically configured for marshalling and unmarshalling of CloudEvents.
+
+With the library imported, you can now update the application to use CloudEvents.
+
+The following sections guide you through updating the `SystemResource` class from the Getting Started guide  to use CloudEvents. You can view the <<completed_class>> to check that changes you make align with the expected result.
+
+==== Returning a CloudEvent
+
+First, update the `SystemResource` class response Type from `Response` to `CloudEvent`. In the method declaration, replace `Response` with `CloudEvent`:
+
+For the Method Declaration replace `Response` with `CloudEvent`
+```
+public CloudEvent getProperties() {
+```
+
+Now need to construct a CloudEvent to return. However, first we need to do some work on the system properties to be able to include them as the data within the event.
+
+CloudEvents cannot convert the data from Object to `byte[]` and requires the data to be in a binary format when it is provided during its building process. As such, we can take the properties from `System.getProperties()` and make them into a JSON string by using `Jsonb`.
+```
+/* java.util.properties does not have a direct way to obtain a byte[] so store in an intermediary Map first*/
+    Map properties = System.getProperties();
+    Jsonb jsonb = JsonbBuilder.create();
+/* convert properties map into a JSON string which can then be converted into a byte[]*/
+    String jsonString = jsonb.toJson(properties);
+
+```
+
+With our data in string format, we can now get the `byte[]` representation of the data.
+
+The `CloudEventBuilder` class provides the necessary components to build our CloudEvent.  Use the most recent specification version, which is `v1()`.
+```
+    return CloudEventBuilder.v1()
+        .withData(jsonString.getBytes())
+        .withDataContentType("application/json")
+        .withId("properties")
+        .withType("java.properties")
+        .withSource(URI.create("http://system.poperties"))
+        .build();
+```
+Besides `withData()`, the rest of the methods set the values that will be returned as headers in the response.
+Once all the required properties are set, you can build the CloudEvent Object.
+
+If you are missing any required properties when your application attempts to build the CloudEvent, an exception will be thrown that informs you of a missing property. To see which properties are required, you can review the https://github.com/cloudevents/spec/blob/v1.0/spec.md[specification].
+
```

Comment:
why would they be missing required properties if they are following these steps?

Reply:
This is purely a note to the developer for when they are doing it in their own application what would happen. Would a better way of providing this information be: the Sample above provides only the required properties for a valid CloudEvent. If any of these properties were to be missing then an exception would be thrown

Reply:
updated